services/scabbard/libscabbard/src/store/command/prepare_service.rs

```diff
@@ -12,29 +12,96 @@
+    service::ScabbardService,
+    state::Scabbard2pcState,
+    ScabbardStoreFactory,
+};
 
-#[derive(Default)]
 pub struct ScabbardPrepareServiceCommand<C> {
-    _store_factory: PhantomData<C>,
+    store_factory: Arc<dyn ScabbardStoreFactory<C>>,
+    scabbard_service: ScabbardService,
 }
 
 impl<C> ScabbardPrepareServiceCommand<C> {
-    pub fn new() -> Self {
+    pub fn new(
+        store_factory: Arc<dyn ScabbardStoreFactory<C>>,
+        scabbard_service: ScabbardService,
+    ) -> Self {
         Self {
-            _store_factory: PhantomData,
+            store_factory,
+            scabbard_service,
         }
     }
 }
 
 impl<C> StoreCommand for ScabbardPrepareServiceCommand<C> {
     type Context = C;
 
-    fn execute(&self, _conn: &Self::Context) -> Result<(), InternalError> {
-        info!("executing prepare service command");
+    fn execute(&self, conn: &Self::Context) -> Result<(), InternalError> {
+        let context = create_context(&self.scabbard_service)?;
+
+        let store = self.store_factory.new_store(conn);
+
+        store
+            .add_service(self.scabbard_service.clone())
+            .map_err(|err| InternalError::from_source(Box::new(err)))?;
+
+        store
+            .add_consensus_context(
+                self.scabbard_service.service_id(),
+                ScabbardContext::Scabbard2pcContext(context),
+            )
+            .map_err(|err| InternalError::from_source(Box::new(err)))?;
         Ok(())
     }
 }
+
+fn create_context(service: &ScabbardService) -> Result<Context, InternalError> {
+    let mut peers = service.peers().to_vec();
+
+    peers.push(service.service_id().service_id().clone());
+
+    let coordinator = get_coordinator(peers);
+
+    if service.service_id().service_id() == &coordinator {
+        ContextBuilder::default()
+            .with_coordinator(&coordinator)
+            .with_epoch(1)
+            .with_participants(
+                service
+                    .peers()
+                    .iter()
+                    .map(|participant| Participant {
+                        process: participant.clone(),
+                        vote: None,
+                    })
+                    .collect(),
+            )
+            .with_state(Scabbard2pcState::WaitingForStart)
+            .with_this_process(service.service_id().clone().service_id())
+            .build()
+            .map_err(|err| InternalError::from_source(Box::new(err)))
+    } else {
+        ContextBuilder::default()
+            .with_coordinator(&coordinator)
+            .with_epoch(1)
+            .with_participant_processes(service.peers().to_vec())
+            .with_state(Scabbard2pcState::WaitingForVoteRequest)
+            .with_this_process(service.service_id().clone().service_id())
+            .build()
+            .map_err(|err| InternalError::from_source(Box::new(err)))
+    }
+}
+
+fn get_coordinator(peers: Vec<ServiceId>) -> ServiceId {
```

Comment:
Add API doc to this function pointing out the significance - that this is the algorithm for determining the coordinator.

Reply:
Added